### PR TITLE
Add support for specifying start height in forked mode

### PR DESCRIPTION
**Pull request by peterargue**
## Description

When forking a live network, it's helpful to be able to choose the block to start from. This allows for testing applications against a specific previous state.

This PR adds a new cli flag `--start-block-height` that allows the user to specify a specific block to start from in the forked network. It also enables register value caching to speed up execution.

Note: This changes the behavior of the current forked implementation. Instead of always using the archive node's latest height, it will start at an initial block then increment the height using the same mechanism as the unforked network. This provides a more intuitive behavior, but more importantly allows for register caching.

______

For contributor use:

- [x] Targeted PR against `master` branch
- [ ] Linked to GitHub issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-emulator/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation
- [x] Re-reviewed `Files changed` in the GitHub PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -69,7 +69,8 @@ and if you plan to run the emulator with Docker you must use the environment var
...
 | `--chain-id`                  | `FLOW_CHAINID`               | `emulator`     | Chain to simulate, if 'mainnet' or 'testnet' values are used, you will be able to run transactions against that network and a local fork will be created..  Valid values are: 'emulator', 'testnet', 'mainnet'                                       |
-| `-redis-url`                  | `FLOW_REDIS_URL`             | ''             | Redis-server URL for persisting redis storage backend ( `redis://[[username:]password@]host[:port][/database]` )                                                                                                                                    |
+| `--redis-url`                 | `FLOW_REDIS_URL`             | ''             | Redis-server URL for persisting redis storage backend ( `redis://[[username:]password@]host[:port][/database]` )                                                                                                                                    |
+| `--start-block-height`        | `FLOW_STARTBLOCKHEIGHT`             | `0`             | Start block height to use when starting the network using 'testnet' or 'mainnet' as the chain-id    |
```
Comment (peterargue) [inline README.md:73]: not sure if this is the correct envvar name. I couldn't find them defined anywhere

File: README.md [hunk 2]
```diff
@@ -137,6 +138,9 @@ POST http://localhost:8080/emulator/rollback
 Post Data: height={block height}
 ```
 
+Note: it is only possible to roll back state to a height that was previously executed by the emulator.
+To roll back to a past block height when using a forked mainnet or testnet network, use the
```
Comment (turbolent) [inline README.md:142]:
```suggestion
To roll back to a past block height when using a forked Mainnet or Testnet network, use the
```

File: cmd/emulator/start/start.go [hunk 3]
```diff
@@ -73,6 +73,7 @@ type Config struct {
 	RedisURL                 string        `default:"" flag:"redis-url" info:"redis-server URL for persisting redis storage backend ( redis://[[username:]password@]host[:port][/database] ) "`
 	SqliteURL                string        `default:"" flag:"sqlite-url" info:"sqlite db URL for persisting sqlite storage backend "`
 	CoverageReportingEnabled bool          `default:"false" flag:"coverage-reporting" info:"enable Cadence code coverage reporting"`
+	StartBlockHeight         uint64        `default:"0" flag:"start-block-height" info:"block height to start the emulator at. only valid when forking mainnet or testnet"`
```
Comment (turbolent) [inline cmd/emulator/start/start.go:76]:
```suggestion
	StartBlockHeight         uint64        `default:"0" flag:"start-block-height" info:"block height to start the emulator at. only valid when forking Mainnet or Testnet"`
```

File: cmd/emulator/start/start.go [hunk 4]
```diff
@@ -130,6 +131,10 @@ func Cmd(getServiceKey serviceKeyFunc) *cobra.Command {
 				Exit(1, err.Error())
 			}
 
+			if conf.StartBlockHeight > 0 && flowChainID != flowgo.Mainnet && flowChainID != flowgo.Testnet {
+				Exit(1, "❗  --start-block-height is only valid when forking mainnet or testnet")
```
Comment (turbolent) [inline cmd/emulator/start/start.go:135]:
```suggestion
				Exit(1, "❗  --start-block-height is only valid when forking Mainnet or Testnet")
```

File: docs/overview.md [hunk 5]
```diff
@@ -137,6 +138,9 @@ POST http://localhost:8080/emulator/rollback
 Post Data: height={block height}
 ```
 
+Note: it is only possible to roll back state to a height that was previously executed by the emulator.
+To roll back to a past block height when using a forked mainnet or testnet network, use the
```
Comment (turbolent) [inline docs/overview.md:142]:
```suggestion
To roll back to a past block height when using a forked Mainnet or Testnet network, use the
```

File: docs/overview.md [hunk 6]
```diff
@@ -207,6 +211,18 @@ To generate a service key, use the `keys generate` command in the Flow CLI.
 flow keys generate
 ```
 
+## Emulating mainnet and testnet transactions
+The emulator allows you to simulate the execution of transactions as if they were
+performed on the mainnet or testnet. In order to activate this feature,
```
Comment (turbolent) [inline docs/overview.md:216]:
```suggestion
## Emulating Mainnet and Testnet transactions
The emulator allows you to simulate the execution of transactions as if they were
performed on the Mainnet or Testnet. In order to activate this feature,
```

File: server/server.go [hunk 7]
```diff
@@ -306,10 +308,16 @@ func (s *EmulatorServer) Stop() {
...
 
 	if conf.ChainID == flowgo.Testnet || conf.ChainID == flowgo.Mainnet {
-		storageProvider, err = remote.New(remote.WithChainID(conf.ChainID))
+		provider, err := remote.New(remote.WithChainID(conf.ChainID))
 		if err != nil {
 			return nil, err
 		}
+
+		if conf.StartBlockHeight > 0 {
+			provider.SetBlockHeight(conf.StartBlockHeight)
```
Comment (turbolent) [inline server/server.go:317]:
```suggestion
			err = provider.SetBlockHeight(conf.StartBlockHeight)
			if err != nil {
				return nil, err
			}
```

**Review by turbolent — APPROVED** [review]
Nice!

**Review by turbolent — APPROVED** [review]
LGTM! Not very familiar with storage / versioning, but looks good from a Go perspective

File: README.md (hunk 1 above)
Comment (bluesign) [reply README.md:73]: @peterargue  I feel like it should be `FLOW_START_BLOCK_HEIGHT` 

it should be in "https://github.com/psiemens/sconfig/blob/893e07c4e321243de5ca6596cdb74fbe9046e322/sconfig.go#L162" 

default replace of viper is "-" to "_" afaik 

I will check all flags and go all over them

**Pull request merged**